### ISSUE TITLE
Input (typing) latency regression 0.12.2 -> 0.13.1

### ISSUE THREAD

**Issue by dimhatz** [neovide/neovide#2602]
First of all, thank you very much for this excellent neovim GUI! 

Measured on Windows 10, Ryzen 5600X, Radeon 5500XT 8GB, using typometer v1.1.0 from here: https://github.com/frarees/typometer/releases/tag/v1.1.0
(and oracle java 22 SE)

Command: `java --add-opens=java.desktop/java.awt=ALL-UNNAMED -jar typometer-1.1.0.jar`

Settings as shown in screenshot (chars was limited to 64 due to typometer unable to track further neovide v0.13.1, likely due to fonts being rendered differently). I took multiple measurements to compensate for less chars. 

Starting nvim (v0.10 stable) config:
```lua
vim.opt.number = true
vim.opt.mouse = 'a'
vim.opt.linespace = -1
vim.g.neovide_cursor_animate_in_insert_mode = false -- <-- to ensure minimal latency possible
```

Neovide config.toml:
```toml
size = 11 # <-- 11 for v0.13.1, 15 for v0.12.2
hinting = "full"
edging = "antialias"
normal = { family = "IosevkaCustom NF", style = "W350" } # 350 = SemiLight
bold = { family = "IosevkaCustom NF", style = "W600" } # 600 = Bold
italic = { family = "IosevkaCustom NF", style = "W350 Italic" }
bold_italic = { family = "IosevkaCustom NF", style = "W600 Italic" }
```

Font size was adjusted to make sure we have the same amount of text (lines) before running: 116 lines on my machine when maximized.

The initial nvim state is just a single blank buffer (see `fulltext` annotation below).

**Annotations on measurements**
`gl` -> used `--opengl` for v0.13.1
`dx` -> direct3d, did not use `--opengl` for v0.13.1
`fulltext` -> before taking measurements, the first 115 lines are filled with text (as opposed to a completely empty buffer otherwise), The last line (116) is empty and the cursor is placed there (in insert).

**TLDR / Findings**
With release 0.13.1 (and 0.13.0) the input latency is `>2x` vs 0.12.2.

d3d performs around 25% worse than opengl, at least on this machine. Additionally, with d3d there is some lag when scrolling, which comes and goes. During these laggy states, the gpu usage spikes ~35% -> ~55%. With opengl, gpu usage is stable at ~30% (when not idle).

When using `--no-vsync --no-idle` and ` vim.g.neovide_refresh_rate = 60`, with opengl, the latency improves by ~25%,
with d3d worsens 3x (!).

When using `--no-multigrid`, the latency improves across the board, but the best results are still with v0.12.2, when the screen is full
of text, see entries annotated with `fulltext`. With a completely empty buffer v0.12.2 and v0.13.1 opengl are the same.

![typometer](https://github.com/neovide/neovide/assets/7132507/1adccaad-b105-4132-8fb8-275604a6c9a5)

**Comment by xzbdmw** [comment]
I can't use typometer's Native api at macos, the async mode shows avg 20ms in 12.1 and 23s in 13.1, not sure it is accurate or not.

**Comment by dimhatz** [comment]
@xzbdmw `async` mode (setting) in typometer should be disabled for more accurate results, it only affects the way chars on screen are recognized.

The vsync setting (I assume in MacOs settings), when enabled, with 60hz refresh rate, is expected to give an increase of 8.3ms on average (due to quantization). AFAIU, the time is rounded up towards the nearest frame, which lasts for 16.6ms on 60hz. If the actual times are evenly distributed, the average increase is expected to be half the frame, that is, 8.3ms. So, the extra 10ms that you see with vsync on is normal.

**Comment by fredizzimo** [comment]
I can replicate the same results with https://game.intel.com/us/intel-presentmon/, which is far more reliable. 

However, I don't think there's much that can be done about it, the lag consist of the following
1 ms - the key press is sent to Neovim, processed there and back. This can take much longer depending on which plugins you have installed though. But outside Neovide's control
0-1 frame (16.6/ ms in this case) or 8.3 ms  in average to wait until we are allowed to render. This is optimized away for the first keypress after the rendering has been idle. So, in that case you have only 1ms latency.
2 frames ( 33.33 ms) of latency caused by Windows composition. This can be optimized if you run `g:neovide_fullscreen=1` and run Windows 11 with at least the 22h2 patch with optimized borderless fullscreen rendering.

So, in total you have around 34 ms in the best case, which is also the case where the latency matters the most, the initial keypress of something. And the worst case around 50 ms.

The novsync mode is not properly implemented for D3D , so that's why you see strange results with that. And opengl did not properly sync with the compositor, which explains why it has a lower latency. I'm don't know what's changed in the 13.x though, but I have observed that opengl sometimes syncs correctly, and that's probably why you saw the around 34 ms delay.

For more information, see also this article https://jackmin.home.blog/2018/12/14/swapchains-present-and-present-latency/

@xzbdmw, the macOS has some strange synchronization problems, and I believe this will fix it https://github.com/neovide/neovide/pull/2461

**Comment by xzbdmw** [comment]
Does multigrid introduces latency? If so can it be improved?

**Comment by dimhatz** [comment]
@fredizzimo Thank you for sharing the article and the tool link, very useful indeed!

Here are some observations from PresentMon and v0.13.1 (Windows 10, windows maximized but not fullscreen, `--no-idle`, `vim.g.neovide_cursor_animate_in_insert_mode = false`):

- With opengl has sync interval reported as `-1` with an initial spike to `0` during startup. With direct3d it is always at `1` with an initial spike to `2` during startup.
- Direct3d: Just filling the whole screen with chars (every line, no overflowing, about 116 lines) results in spiky frame times with lots of spikes to 32ms and 50 ms. I think adding a new line (to the already filled screen) and typing very fast produces even more of these spikes, but I'm not entirely sure. For opengl the frame times are always at 16ms, no spikes. 
- I am sure I have seen some dropped frames too with directx (during the filled screen test), but cannot reproduce now.
- v0.12.2 shows metrics similar to 0.13.1 with opengl, no frame time spikes.

I hope that helps, let me know if you would like to me to run some more tests.

```
And opengl did not properly sync with the compositor, which explains why it has a lower latency. 
I'm don't know what's changed in the 13.x though, but I have observed that opengl sometimes 
syncs correctly, and that's probably why you saw the around 34 ms delay.
```

You mean v0.12 did not sync correctly, and that is what caused it to have lower latency? If so, maybe this is a feature? :)
AFAICT alacritty uses opengl too and I measure 16ms in nvim (with typometer, but with a smaller window). As opposed to v0.13.1 that is always >32ms no matter the window size. Not sure what vscode's electron uses, but it's also at 16ms (maximized).

I think we should not dismiss typometer just yet, even though its accuracy is worse than PresentMon and similar.
It is still very useful, to detect issues when the frames themselves render perfectly, but their content (produced by nvim) is delayed.
At the end of the day, the user can only see what they have typed when it actually appears, and this is what typometer measures.

P.S. As an aside, I have recently stumbled upon these videos https://www.youtube.com/playlist?list=PLEMXAbCVnmY6zCgpCFlgggRkrp0tpWfrn, where the author explains how he wrote an extremely performant terminal. I think you may find them interesting and useful (in case you have not already seen them).

**Comment by fredizzimo** [comment]
@xzbdmw, no not unless you have some plugin on the Neovim side that slows things down. Or maybe if there's something wrong with the blur/shadow/transparency rendering, but that can only happen when a float is open.

There's also a bug in Neovim, that sends a grid full of garbage if you resize the splits a lot. The splits are rendered on a separate grid, and that is not cleared unless you resize the window. This can slow down the framerates a lot on a big grid. But it clears up as soon as you resize the window. We can and probably should compensate for this in Neovide, but it's not trivial, the actual character grid has to be modified based on the overlap, since the graphics library Skia, which we use does not do it with just clipping rectangles.

@dimhatz, I suspect there's something slowing down the rendering in general for you, can you send a tracy log with typing and scrolling around https://neovide.dev/troubleshooting.html#performance-profiling

**Comment by dimhatz** [comment]
Sure. I do have access to my pc until tomorrow late evening. Will make sure to send it once I have.

**Comment by xzbdmw** [comment]
running with `cargo run --profile profiling --features profiling --
 --no-vsync`
<img width="1762" alt="截屏2024-06-08 04 28 21" src="https://github.com/neovide/neovide/assets/97848247/730d9c65-61ed-4385-90d2-419181f50c92">

running with `cargo run --profile profiling --features profiling --
 --no-vsync --no-multigrid`
<img width="1762" alt="截屏2024-06-08 04 28 37" src="https://github.com/neovide/neovide/assets/97848247/e168c32f-1bb7-4371-8fd8-bc13d165e003">

They looks quite different, if multigrid enanled, those render duration time is 5-10ms, while if mutigrid disables, they are 2ms typically, The context is I repeat delete chars in normal buffer(with real code full screen).

[trace-no-multigrid3.log.zip](https://github.com/user-attachments/files/15744682/trace-no-multigrid3.log.zip)
[trace-multigrid3.log.zip](https://github.com/user-attachments/files/15744685/trace-multigrid3.log.zip)
I'm using my own config, but shouldn't they behave the same with the same config  regards to render speed.

**Comment by xzbdmw** [comment]
With a clean environment, running with 
`cargo run --profile profiling --features profiling --
 --no-vsync -- -u ~/.config/nvim/neovide.lua ~/.config/nvim/neovide.lua`
 and 
 `cargo run --profile profiling --features profiling --
 --no-vsync --no-multigrid -- -u ~/.config/nvim/neovide.lua ~/.config/nvim/neovide.lua`.
 
 multigrid version is still bad compared to no-multigrid version (10ms vs 2ms)
 
 
[trace-no-multigrid-clean.log.zip](https://github.com/user-attachments/files/15744962/trace-no-multigrid-clean.log.zip)

[trace-multigrid-clean.log.zip](https://github.com/user-attachments/files/15744961/trace-multigrid-clean.log.zip)

**Comment by fredizzimo** [comment]
@xzbdmw, the long "frame" you see there is in fact several frames if you zoom in. The reason why it appears to do that is that there's no animation in no-multigrid, so it (almost) always only draws one frame in a row.

For example, zoomed in on the multigrid trace
![image](https://github.com/neovide/neovide/assets/9946255/c28c4de9-eccb-43ba-afce-34eab2b39004)

8 frames are draw in that span, and the latency highlighted in red. You can also see that the vsync is not working as it should, the present times are quite random. But there's some attempt to sync, since `swap_buffers` takes some time. But the actual vsync wait appears to always succeed immediately for you, if zoomed in just before `swap_buffers`
![image](https://github.com/neovide/neovide/assets/9946255/c285bd83-fb3a-4069-b739-8c43273e586d)
The metal renderer https://github.com/neovide/neovide/pull/2461 should fix that. But it's not necessarily a problem until you render enough frames in a row to fill up all the buffers (typically two or three frames). You can check if it starts to block on vsync when you are scrolling.

There are also 4 keyboard events in that span, I believe two presses and two key ups, so that's another reason why it might look like a single frame if not zoomed in enough.

Unfortunately, today will be the last day for me, and even that with quite limited time to look at the logs. After that I will be on vacation with only an Ipad available, so no possibility to look at the logs or make fixes. So, if you have a chance to send the logs, today @dimhatz, it would be good.

**Comment by xzbdmw** [comment]
Yeah but why multigrid disabled version’s swap_buffers succeed within 1ms, they are both run with —no-vsync.

**Comment by fredizzimo** [comment]
You can easily find examples of it taking longer, for example here, from the `no-multigrid-clean` log.
https://github.com/neovide/neovide/issues/2602#issuecomment-2156022241

There's one thing I don't understand though, you say that you run with `--no-vsync` and with a `vim.g.neovide_refresh_rate = 60`. That should use a software based timer https://github.com/neovide/neovide/blob/1db63d93b18cc1eb5ac5b3e67b3ca63505f10800/src/renderer/vsync/vsync_timer.rs#L19 and never let the swaps be closer than 16.66 ms from each other. But that's not what's happening on your system. On my Linux system that works as expected though, although it's very choppy due to the lack of vsync as expected.

**Comment by dimhatz** [comment]
@fredizzimo Sorry for the delay, just got back!

Here are the logs:

[empty-buf.log](https://github.com/user-attachments/files/15748266/empty-buf.log)
[empty-buf-0.12.2.log](https://github.com/user-attachments/files/15748267/empty-buf-0.12.2.log)

[filling-buf.log](https://github.com/user-attachments/files/15748268/filling-buf.log)
[filling-buf-0.12.2.log](https://github.com/user-attachments/files/15748270/filling-buf-0.12.2.log)

[full-buf.log](https://github.com/user-attachments/files/15748271/full-buf.log)
[full-buf-0.12.2.log](https://github.com/user-attachments/files/15748272/full-buf-0.12.2.log)

[scrolling.log](https://github.com/user-attachments/files/15748273/scrolling.log)
[scrolling-0.12.2.log](https://github.com/user-attachments/files/15748274/scrolling-0.12.2.log)

The main logs are for 0.13.1 tag (I can also run these for the latest if needed)
The logs for 0.12.2 are annotated accordingly.
The 0.13.1 version was run with `--opengl`.
Both versions had `vim.g.neovide_cursor_animate_in_insert_mode = false`, fonts were adjusted to produce the same amount of text.

- `empty-buf.log`: typometer types 30 chars into an empty buffer (measures 32ms for 0.13.1, at 16ms for 0.12.2)
- `filling-buf.log`: pasting a full line of text (just dots) until the screen is full by holding `p`. Both versions have slowdowns: when I let go of `p`, the lines continue to be added. Note: I have a high keyboard repeat rate, but still..
- `full-buf.log`: just typing random stuff on a new line after the screen was filled (as above)
- `scrolling.log`: scrolling 1000 full lines (as above) back and forth - here is where I cannot see any difference between the versions, adding just for reference.

P.S. Have a nice vacation! Please take your time to rest!

I will be here when you get back and have time again, no worries. I will try to bisect and maybe find the offending commit.

Thanks again for your good work!

**Comment by fredizzimo** [comment]
From Neovide's perspective everything looks normal except for the `filling-buf` (more about that later), for example from `full-buf`
![image](https://github.com/neovide/neovide/assets/9946255/5edb790d-48f9-4e9a-ab13-65c8f23c3b9e)

A keyboard event is received, we wait until the vsync, and then render the fame in 1ms, creating a total lag of 10.97ms in this example, but it can range from around 1ms to 17ms. Then there are two frames of additional lag from the compositor as explained in my earlier. post

And now to `filling-buf

It looks like your keyboard repeat rate is set to 16 ms, but Neovim is struggling to keep up, sometimes it takes almost that time for Neovim to just process the event
![image](https://github.com/neovide/neovide/assets/9946255/3d1855da-4d6b-49ea-a64a-3ef4bcd49c18)

But it never seems to take longer than 16.66 ms, so that's not the full explanation. If that happens, then there's indeed big trouble because the `nvim-rs` library we support does not support Neovim rpc notifications, only requests, so we are forced to wait for the response for each event. And if that does not happen in time then the events start buffering up. That's something I have been aware of for a long time and that we should get fixed at some point, since it also affects network links where the latency is bigger than the keyboard repeat rate. See https://github.com/KillTheMule/nvim-rs/issues/7

The real problem seems to be that we receive updates after Neovide has stopped sending new input to Neovim:
![image](https://github.com/neovide/neovide/assets/9946255/c3aef078-41c7-4e7b-a2b4-a5680d588bbc)

So Neovim is doing something here, perhaps running the clipboard provider https://neovim.io/doc/user/provider.html#provider-clipboard. 

So, at least based on this quick analysis the problem there seem to be on the Neovim side, and further profiling is needed for that.

**Comment by dimhatz** [comment]
Indeed, it seems that this only happens when pasting. If I insert a line full of chars (with a line break) and then press `.`, I get the same chars in the buffer, but no lag.

This pasting scenario from (my) user standpoint is a minor issue, I only mentioned it in case it was at all related to the latency.

What I cannot get is why v0.12.2 manages to put the "user-perceivable pixels" on the screen at 16ms, vs 32ms of v0.13.1.

Does the v0.12.2 code call the drawing APIs differently?

If both versions produce the drawable result in the same time (e.g. in the simplest `empty-buf` example), how does v0.12.2 avoids the extra 16ms of sync penalty?

**Comment by dimhatz** [comment]
Maybe the following provides a clue:

- With small neovide v0.13.1 window (31 lines, 80 columns), typometer latency is at ~16ms.
- With maximized (but not fullscreen) neovide window (116 lines, 412 columns) typometer latency is at ~35ms - not a single frame under 32ms.
(In v0.12.2 both cases are at ~16ms)

![typometer-small-win-big-win](https://github.com/neovide/neovide/assets/7132507/1934bb9a-67c6-4553-a8bd-e2b0d67abcc8)


Log for both sequences (first small window + 30 char typed, then maximized + 30 chars typed)

[small-win-big-win.log](https://github.com/user-attachments/files/15749063/small-win-big-win.log)

**Comment by dimhatz** [comment]
Bisect result:

```
278bf6bfe596ce1dc786f98ae780203739f95210 is the first bad commit
commit 278bf6bfe596ce1dc786f98ae780203739f95210
Author: Kaylee Simmons <kay@the-simmons.net>
Date:   Fri Apr 19 13:34:31 2024 -0700

    feat: render same z-index together (#2467)

    * feat: render same z-index together

    * make clippy happy

    * uniform blends in layer, render all background first, then foreground

    * fix background blends

    * add test for telescope cases

    * fix telescope case

    * fix test case

    * Remove custom hull algorithm in favor of skia's built in one

    * Remove zindex step setting and use consecutive groupings of floating windows instead

    * Swap to iter_lines

    * Add back in the floating order which was removed due to a clippy warning but is now needed again

    * Address clippy issues and refactor rendered window lines functions to split out their components into parts

    * Fix overzealous clipping

    ---------

    Co-authored-by: Hawtian Wang <twistoy.wang@gmail.com>
    Co-authored-by: Hawtian Wang <hawtian.wang@twistoy.cn>

 .gitignore                      |   5 +
 Cargo.lock                      |   7 +
 Cargo.toml                      |   1 +
 src/renderer/mod.rs             |  79 +++++++++-
 src/renderer/rendered_layer.rs  | 224 +++++++++++++++++++++++++++
 src/renderer/rendered_window.rs | 326 ++++++++++++++++++++++++----------------
 src/units.rs                    |  10 ++
 src/utils/mod.rs                |   4 +
 src/utils/test.rs               | 218 +++++++++++++++++++++++++++
 9 files changed, 734 insertions(+), 140 deletions(-)
 create mode 100644 src/renderer/rendered_layer.rs
 create mode 100644 src/utils/test.rs
```


Enjoy your vacation @fredizzimo!
Hopefully we get to the bottom of this.

**Comment by dimhatz** [comment]
The latest main did not fix the issue for me. Also, Radeon GPU Profiler seems to only support dx12 and vulkan, so it's a no-go for opengl. Thanks for the suggestion though.

Since the issue, as you said, is likely on the GPU side, there is not much we can do further.

I have updated the drivers to latest, tried various driver settings (forcing vsync on / off, "enhanced sync", "anti-lag", opengl triple buffering) but unfortunately none made a difference.

When I get my hands on my friend's linux machine (intel gpu), I will run latency tests to further confirm this is only limited to AMD cards. Will post the results here.

Thank you for all your help!

**Comment by fredizzimo** [comment]
I'm not that interested in knowing what Opengl does, it seems to bypass the compositor at least partially, which can reduce the lag but at the cost of dropped frames and stuttering. So profiling the current D3D implementation on master should be enough.

For the record when I profiled this on my Nvidia card quickly before I left, I found no applications that had less than 2 frames of latency. 2 frames of  latency is the normal minimum latency on Windows. The results might have been slightly different without the presentmon overlay enabled, see the paragraph below.

The difference you see with a small and big window could be explained by the multiplane overlay optimization that is sometimes applied though https://james.darpinian.com/blog/latency-platform-considerations, https://learn.microsoft.com/en-us/windows-hardware/drivers/display/multiplane-overlay-hardware-requirements. It's also possible that the support for transparency totally disables that optimization https://github.com/neovide/neovide/pull/2529.

As mentioned earlier, we can get it closer to this theoretical 2 frames minimum by skipping the initial vsync wait.  That optimization was removed here, because it's tricky to implement properly https://github.com/neovide/neovide/pull/2359. So, that's probably something we should try to optimize.

One thing to note, if you are interested in low input latency it's worth considering a high refresh monitor, for example 144hz gives you around 20ms of latency currently.

**Comment by dimhatz** [comment]
Thanks for the informative articles! It may indeed be the case that on my system "Multiplane Overlay" is active. I could not confirm this, since "Hardware Composed: Independent Flip" widget in PresentMon seems to be missing for me, but getting 16ms typometer latencies would be impossible otherwise AFAICT.

I have run a few typometer tests with D3D from latest main.

Interestingly, with D3D, when the window is about 2/3 (3/4?) of total screen size the measured latency is ~16ms. When it is more than that (even when not maximized, just manually resized), the latency is ~43ms. Maybe Windows' dwm is doing something weird, like disabling multiplane overlay for a window of a certain size.

Tracy profiles seem to be the same (to my untrained eye):

[d3d-smaller-16ms.log](https://github.com/user-attachments/files/15839726/d3d-smaller-16ms.log)
[d3d-bigger-42ms.log](https://github.com/user-attachments/files/15839733/d3d-bigger-42ms.log)

Regarding monitor, good advice, but unfortunately I will not be able to get a high-refresh rate one in the foreseeable future.

I will install linux on an old hdd, to see if things are better there. Will report the findings here. Maybe this is the time to switch again, now that the whole wayland thing seems to be stabilizing (I was meaning to do it for quite some time, but kept delaying, since previously there was no way to get custom mouse gestures a-la Easystroke under wayland).

**Comment by dimhatz** [comment]
Finally got some time to run some tests on linux. Same hardware (Ryzen 5 5600, Radeon 5500XT, amdgpu driver - default), KDE Neon, X11, compositing disabled.

```lua
vim.g.neovide_cursor_animate_in_insert_mode = false
vim.g.neovide_refresh_rate = 1000
```

1 default (no additional cmdline args) -> 28ms input latency.
2. `--no-vsync --no-idle` -> ~10ms - the fastest
3. just `--no-vsync` ->  ~17ms

Could not take measurements under wayland, since typometer does not work with it. Not sure if currently there is a such a tool for wayland, since screen capturing is supposed to be forbidden (except for special compositor-specific apps?).

Tracy logs attached if anyone is interested to take a look.

[1-linux-default.log](https://github.com/user-attachments/files/16180138/1-linux-default.log)
[2-linux-noidle-novsync.log](https://github.com/user-attachments/files/16180141/2-linux-noidle-novsync.log)
[3-linux-novsync.log](https://github.com/user-attachments/files/16180144/3-linux-novsync.log)

P.S. `--no-multigrid` does not seem to affect measurements.

**Comment by tfriedel** [comment]
> For the record when I profiled this on my Nvidia card quickly before I left, I found no applications that had less than 2 frames of latency. 2 frames of latency is the normal minimum latency on Windows. 

I have a 60hz monitor and on Win10 there's plenty of terminals or editors (e.g. notepad) where I measure 1 frame / 16ms.
On Windows 11 on the other hand I observed what you describe: at least 2 frames lag with nvidia cards. 

You may be interested in these efforts to reduce input latency for VSCode:
https://github.com/microsoft/vscode/issues/221145
